lib/SILOptimizer/Transforms/SpecializeOpaqueArchetypes.cpp

```diff
@@ -0,0 +1,466 @@
+  }
+
+  void visitYieldInst(YieldInst *Inst) {
+    auto OrigValues = Inst->getYieldedValues();
+    auto Values = getOpValueArray<8>(Inst->getYieldedValues());
+    auto ResumeBB = getOpBasicBlock(Inst->getResumeBB());
+    auto UnwindBB = getOpBasicBlock(Inst->getUnwindBB());
+    for (auto idx : indices(Values)) {
+      if (OrigValues[idx]->getType().getASTType() !=
+          Values[idx]->getType().getASTType()) {
+        Values[idx] = createCast(RegularLocation::getAutoGeneratedLocation(),
+                                 Values[idx], OrigValues[idx]->getType());
+      }
+    }
+
+    getBuilder().setCurrentDebugScope(getOpScope(Inst->getDebugScope()));
+    recordClonedInstruction(
+        Inst, getBuilder().createYield(getOpLocation(Inst->getLoc()), Values,
+                                       ResumeBB, UnwindBB));
+  }
+
+  void visitCopyAddrInst(CopyAddrInst *Inst) {
+    auto src = getOpValue(Inst->getSrc());
+    auto dst = getOpValue(Inst->getDest());
+    auto srcType = src->getType();
+    auto destType = dst->getType();
+    getBuilder().setCurrentDebugScope(getOpScope(Inst->getDebugScope()));
+    // If the types mismatch cast the operands to the non opaque archetype.
+    if (destType.getASTType() != srcType.getASTType()) {
+      if (srcType.getASTType()->hasOpaqueArchetype()) {
+        src = getBuilder().createUncheckedAddrCast(
+            getOpLocation(Inst->getLoc()), src, destType);
+      } else if (destType.getASTType()->hasOpaqueArchetype()) {
+        dst = getBuilder().createUncheckedAddrCast(
+            getOpLocation(Inst->getLoc()), dst, srcType);
+      }
+    }
+    recordClonedInstruction(
+        Inst, getBuilder().createCopyAddr(getOpLocation(Inst->getLoc()), src,
+                                          dst, Inst->isTakeOfSrc(),
+                                          Inst->isInitializationOfDest()));
+  }
+
+  void visitStoreInst(StoreInst *Inst) {
+    auto src = getOpValue(Inst->getSrc());
+    auto dst = getOpValue(Inst->getDest());
+    auto srcType = src->getType();
+    auto destType = dst->getType();
+    getBuilder().setCurrentDebugScope(getOpScope(Inst->getDebugScope()));
+    // If the types mismatch cast the operands to the non opaque archetype.
+    if (destType.getASTType() != srcType.getASTType()) {
+      if (srcType.getASTType()->hasOpaqueArchetype()) {
+        assert(!srcType.isAddress());
+        src = createCast(getOpLocation(Inst->getLoc()), src,
+                         destType.getObjectType());
+      } else if (destType.getASTType()->hasOpaqueArchetype()) {
+        dst = getBuilder().createUncheckedAddrCast(
+            getOpLocation(Inst->getLoc()), dst, srcType.getAddressType());
+      }
+    }
+
+    if (!getBuilder().hasOwnership()) {
+      switch (Inst->getOwnershipQualifier()) {
+      case StoreOwnershipQualifier::Assign: {
+        auto *li = getBuilder().createLoad(getOpLocation(Inst->getLoc()), dst,
```

Comment:
Why do you lower assigns to load/store here instead of letting them be?

Reply:
Same code as SILCloner.

Reply:
Why does it need to appear here?